alf/layers.py

```diff
@@ -110,6 +110,112 @@ def forward(self, input):
             input, num_classes=self._num_classes).to(torch.float32)
 
 
+@gin.configurable
+class FixedDecodingLayer(nn.Module):
+    def __init__(
+            self,
+            input_size,
+            output_size,
+            basis_type="rbf",
+            min_v=-1.,
+            max_v=1.,
+            sigma=None,
+    ):
+        """A layer that uses a set of fixed basis for decoding the inputs.
+        Args:
+            input_size (int): the size of input to be decoded, representing the
+                number of representation coefficients
+            output_size (int): the size of the decoded output
+            basis_type (str): the type of basis to be used for decoding
+                - "poly": polynomials basis
+                - "rbf": radial basis functions
+                - "haar": haar wavelet basis
+            min_v (float): lower bound of the support for the input
+            max_v (float): upper bound of the support for the input
+            sigma (float): the bandwidth parameter used for RBF basis. If None,
+                a default value of ``sigma = (max_v - min_v) / output_size``
+                will be used.
+        """
+        # get the argument list with vals
+        self._kwargs = copy.deepcopy(locals())
+        self._kwargs.pop('self')
+        self._kwargs.pop('__class__')
+
+        super(FixedDecodingLayer, self).__init__()
+
+        assert input_size > 0, "input_size should be at least one"
+        assert basis_type in {"poly", "rbf", "haar"
+                              }, ("the specified method "
+                                  "{} is not supported".format(basis_type))
+
+        self._B = nn.Linear(input_size, output_size, bias=False)
+
+        def _poly_matrix(n, D):
+            # non-square matrix [n, D + 1]
+            x = torch.linspace(min_v, max_v, n)
```

Comment:
we don't need min_v, max_v. Can simply assume min_v =-1, max_v=1. And it may be better to scale B[:,d]  by 1/n^2 or 1/2^n which encode our prior knowledge of preferring lower order polynomial. And we can do similar scaling for haar basis.

Reply:
Removed min_v and max_v. Added basis scaling according to frequencies.